pkg/reconciler/common/prune.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	batchv1 "k8s.io/api/batch/v1"
+	v1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	openshiftPrefix  = "openshift"
+	kubePrefix       = "kube"
+	tektonSA         = "tekton-pipelines-controller"
+	CronName         = "resource-pruner"
+	JobsTKNImageName = "IMAGE_JOB_PRUNER_TKN"
+	defaultTKNImage  = "gcr.io/tekton-releases/dogfooding/tkn@sha256:f69a02ef099d8915e9e4ea1b74e43b7a9309fc97cf23cb457ebf191e73491677"
+)
+
+func Prune(k kubernetes.Interface, ctx context.Context, pru v1alpha1.Prune, targetNamespace string) error {
+	logger := logging.FromContext(ctx)
+	pruningNamespaces, err := GetPrunableNamespaces(k, ctx)
+	if err != nil {
+		return err
+	}
+	if err := createCronJob(k, ctx, pru, targetNamespace, pruningNamespaces); err != nil {
+		logger.Error("failed to create cronjob ", err)
+	}
+
+	return nil
+}
+
+func GetPrunableNamespaces(k kubernetes.Interface, ctx context.Context) ([]string, error) {
+	nsList, err := k.CoreV1().Namespaces().List(ctx, v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var allNameSpaces []string
+	for _, ns := range nsList.Items {
+		if !strings.HasPrefix(ns.Name, openshiftPrefix) && !strings.HasPrefix(ns.Name, kubePrefix) {
+			allNameSpaces = append(allNameSpaces, ns.Name)
+		}
+	}
+	return allNameSpaces, nil
+}
+
+func createCronJob(k kubernetes.Interface, ctx context.Context, pru v1alpha1.Prune, targetNs string, pruningNs []string) error {
+	pruneContainers := getPruningContainers(pru.Resources, pruningNs, pru.Keep)
+	backOffLimit := int32(3)
+	ttlSeconfsAfterFindished := int32(3600)
+	cj := &v1beta1.CronJob{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "CronJob",
+			APIVersion: "batch/v1beta1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: CronName,
+		},
```

Comment:
we need to add an ownerReference to TektonConfig instance, so that this cron job is cascade deleted when we delete the TektonConfig instance.